src/librustc/ty/layout.rs

```diff
@@ -648,6 +648,25 @@ impl Scalar {
             false
         }
     }
+
+    /// Returns a range suitable to be passed to LLVM for range metadata.
+    pub fn range_metadata<C: HasDataLayout>(&self, cx: C) -> Option<Range<u128>> {
+        // For a (max) value of -1, max will be `-1 as usize`, which overflows.
+        // However, that is fine here (it would still represent the full range),
+        // i.e., if the range is everything.
+        let bits = self.value.size(cx).bits();
+        assert!(bits <= 128);
+        let mask = !0u128 >> (128 - bits);
+        let start = self.valid_range.start;
+        let end = self.valid_range.end.wrapping_add(1);
```

Comment:
Oh, wait, this is turning it into an exclusive range. Whereas the storage uses inclusive. I wonder if we could store exclusive ranges throughout, but that sucks for wrap-around ranges because you end up with `0..0` - we should *really* have some wrap-around range type for layout specifically.